Region/BasicStatsCalculator.tcc

```diff
@@ -0,0 +1,100 @@
+#ifndef CARTA_BACKEND_REGION_BASICSTATSCALCULATOR_TCC_
+#define CARTA_BACKEND_REGION_BASICSTATSCALCULATOR_TCC_
+
+#include "BasicStatsCalculator.h"
+
+namespace carta {
+
+template<typename T>
+void BasicStats<T>::join(BasicStats<T>& other) {
+    if (other.num_pixels) {
+        sum += other.sum;
+        sumSq += other.sumSq;
+        num_pixels += other.num_pixels;
+        min_val = std::min(min_val, other.min_val);
+        max_val = std::max(max_val, other.max_val);
+        mean = sum / num_pixels;
+        stdDev = sqrt(sumSq / num_pixels - mean * mean);
+        rms = sqrt(sumSq / num_pixels);
+    }
+}
+
+template<typename T>
+BasicStats<T>::BasicStats(size_t num_pixels, double sum, double mean, double stdDev, T min_val, T max_val, double rms, double sumSq)
+    : num_pixels(num_pixels), sum(sum), mean(mean), stdDev(stdDev), min_val(min_val), max_val(max_val), rms(rms), sumSq(sumSq){
+}
+template<typename T>
+BasicStats<T>::BasicStats()
+    : num_pixels(0), sum(0), mean(0), stdDev(0), min_val(std::numeric_limits<T>::max()), max_val(std::numeric_limits<T>::lowest()), rms(0), sumSq(0) {
+}
+
+template<typename T>
+BasicStatsCalculator<T>::BasicStatsCalculator(const std::vector<T>& data)
+    : _min_val(std::numeric_limits<T>::max()), _max_val(std::numeric_limits<T>::lowest()), _sum(0), _sum_squares(0), _num_pixels(0), _data(data) {}
+
+template<typename T>
+BasicStatsCalculator<T>::BasicStatsCalculator(BasicStatsCalculator<T>& mm, tbb::split)
+    : _min_val(std::numeric_limits<T>::max()), _max_val(std::numeric_limits<T>::lowest()), _sum(0), _sum_squares(0), _num_pixels(0), _data(mm._data) {}
+
+template<typename T>
+void BasicStatsCalculator<T>::operator()(const tbb::blocked_range<size_t>& r) {
+    T t_min = _min_val;
+    T t_max = _max_val;
+    for (size_t i = r.begin(); i != r.end(); ++i) {
+        T val = _data[i];
+        if (std::isfinite(val)) {
+            if (val < t_min) {
```

Comment:
This is going to trigger the same bug as the converter: *both* min and max have to be initialised to the first non-nan value found, otherwise under certain contrived circumstances one of them will end up being incorrect.

Reply:
(Actually maybe less contrived in a small region -- because there will be fewer values in total.)

Reply:
Suggestion for a fix: use something similar to the lambda solution from the converter, but to make it threadsafe make the initial lambda the explicit independent min/max check which also sets the lambda to the lazy check. It doesn't matter if multiple threads call it; the result will be the same.

Reply:
(Actually, how is this loop threadsafe in the first place? Will threads not overwrite each other's min/max values? Or am I missing something?)

Reply:
I can't imagine that the lambda approach would really be faster than two comparisons.

The issue in the converter was because we were doing an `if...else`. Here it's an `if...if`, so we don't have the same problem

Reply:
Oh, I missed that. Never mind, then. :)